modules/services/recoll.nix

```diff
@@ -0,0 +1,170 @@
+{ config, options, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.recoll;
+
+  # The key-value generator for Recoll config format. For future references,
+  # see the example configuration from the package (i.e.,
+  # `$out/share/recoll/examples/recoll.conf`).
+  mkRecollConfKeyValue = generators.mkKeyValueDefault {
+    mkValueString = v:
+      if v == true then
+        "1"
+      else if v == false then
+        "0"
+      else
+        generators.mkValueStringDefault { } v;
+  } " = ";
+
+  # A modified version of 'lib.generators.toINI' that also accepts top-level
+  # attributes as non-attrsets.
+  toRecollConf = { listsAsDuplicateKeys ? false }:
+    attr:
+    let
+      toKeyValue = generators.toKeyValue {
+        inherit listsAsDuplicateKeys;
+        mkKeyValue = mkRecollConfKeyValue;
+      };
+      mkSectionName = name: strings.escape [ "[" "]" ] name;
+      config = mapAttrsToList (k: v:
+        if isAttrs v then
+          ''
+            [${mkSectionName k}]
+          '' + toKeyValue v
+        else
+          toKeyValue { "${k}" = v; }) attr;
```

Comment:
Perhaps this produces unexpected output when a top-level attribute name is greater than an non-top-level attribute name? I.e.

``` nix
{
  a = { foo = "bar"; };
  b = 10;
}
```
would, I guess, produce something like
``` ini
[a]
foo = "bar"

b = 10
```
In this case the `b` key in the INI would be interpreted as being part of `[a]`?

To avoid this I imagine one would have to map the attribute set to a list and then sort this so that attributes with non-attribute set values come first.

Reply:
Nice catch! In Recoll, once a section is defined, it seems to consider all value definitions under that section. At least this is how the behavior of the application goes in my tests. I've added this as `recoll-config-format-order` test.

The updated implementation is pretty naive but it pass the tests. My nix-fu is not as good as I thought. ;p